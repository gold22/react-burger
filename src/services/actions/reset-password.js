import {
    SEND_RESET_USER_PASSWORD_EMAIL_REQUEST,
    SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS,
    SEND_RESET_USER_PASSWORD_EMAIL_ERROR,
    RESET_EMAIL_SENT,
    RESET_USER_PASSWORD_REQUEST,
    RESET_USER_PASSWORD_SUCCESS,
    RESET_USER_PASSWORD_ERROR,
    RESET_PASSWORD_RESET,
} from '../constants/reset-password';

export const sendResetUserPasswordEmail = (email) => async (dispatch, getState, apiClient) => {
    dispatch({
        type: SEND_RESET_USER_PASSWORD_EMAIL_REQUEST,
    });
    try {
        await apiClient.sendResetUserPasswordEmail(email);
        dispatch({
            type: SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS,
        });
    } catch (error) {
        dispatch({
            type: SEND_RESET_USER_PASSWORD_EMAIL_ERROR,
            message: error.message,
        });
    }
};

export const resetEmailSent = () => ({
    type: RESET_EMAIL_SENT,
});

export const resetUserPassword = ({ password, token }) => async (dispatch, getState, apiClient) => {
    dispatch({
        type: RESET_USER_PASSWORD_REQUEST,
    });
    try {
        await apiClient.resetUserPassword({ password, token });
        dispatch({
            type: RESET_USER_PASSWORD_SUCCESS,
        });
    } catch (error) {
        dispatch({
            type: RESET_USER_PASSWORD_ERROR,
            message: error.message,
        });
    }
};

export const resetPasswordReset = () => ({
    type: RESET_PASSWORD_RESET,
});
