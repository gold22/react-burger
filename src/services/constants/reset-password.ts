export const SEND_RESET_USER_PASSWORD_EMAIL_REQUEST: 'SEND_RESET_USER_PASSWORD_EMAIL_REQUEST' = 'SEND_RESET_USER_PASSWORD_EMAIL_REQUEST';
export const SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS: 'SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS' = 'SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS';
export const SEND_RESET_USER_PASSWORD_EMAIL_ERROR: 'SEND_RESET_USER_PASSWORD_EMAIL_ERROR' = 'SEND_RESET_USER_PASSWORD_EMAIL_ERROR';
export const RESET_EMAIL_SENT: 'RESET_EMAIL_SENT' = 'RESET_EMAIL_SENT';
export const RESET_USER_PASSWORD_REQUEST: 'RESET_USER_PASSWORD_REQUEST' = 'RESET_USER_PASSWORD_REQUEST';
export const RESET_USER_PASSWORD_SUCCESS: 'RESET_USER_PASSWORD_SUCCESS' = 'RESET_USER_PASSWORD_SUCCESS';
export const RESET_USER_PASSWORD_ERROR: 'RESET_USER_PASSWORD_ERROR' = 'RESET_USER_PASSWORD_ERROR';
export const RESET_PASSWORD_RESET: 'RESET_PASSWORD_RESET' = 'RESET_PASSWORD_RESET';
