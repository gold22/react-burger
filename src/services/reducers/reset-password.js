import {
    SEND_RESET_USER_PASSWORD_EMAIL_REQUEST,
    SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS,
    SEND_RESET_USER_PASSWORD_EMAIL_ERROR,
    RESET_EMAIL_SENT,
    RESET_USER_PASSWORD_REQUEST,
    RESET_USER_PASSWORD_SUCCESS,
    RESET_USER_PASSWORD_ERROR,
    RESET_PASSWORD_RESET,
} from '../constants/reset-password';

const initialState = {
    isEmailSent: false,
    isSendingEmail: false,
    sendingEmailError: null,
    isPasswordReset: false,
    isResettingPassword: false,
    resetPasswordError: null,
};

// eslint-disable-next-line import/prefer-default-export,@typescript-eslint/default-param-last
export const resetPasswordReducer = (state = initialState, action) => {
    switch (action.type) {
    case SEND_RESET_USER_PASSWORD_EMAIL_REQUEST: {
        return {
            ...state,
            isSendingEmail: true,
        };
    }
    case SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS: {
        return {
            ...initialState,
            isEmailSent: true,
        };
    }
    case SEND_RESET_USER_PASSWORD_EMAIL_ERROR: {
        return {
            ...initialState,
            sendingEmailError: action.message,
        };
    }
    case RESET_EMAIL_SENT: {
        return {
            ...initialState,
        };
    }
    case RESET_USER_PASSWORD_REQUEST: {
        return {
            ...state,
            isResettingPassword: true,
        };
    }
    case RESET_USER_PASSWORD_SUCCESS: {
        return {
            ...initialState,
            isPasswordReset: true,
        };
    }
    case RESET_USER_PASSWORD_ERROR: {
        return {
            ...initialState,
            resetPasswordError: action.message,
        };
    }
    case RESET_PASSWORD_RESET: {
        return {
            ...initialState,
        };
    }
    default: {
        return state;
    }
    }
};
