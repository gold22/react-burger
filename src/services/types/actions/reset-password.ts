import {
    SEND_RESET_USER_PASSWORD_EMAIL_REQUEST,
    SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS,
    SEND_RESET_USER_PASSWORD_EMAIL_ERROR,
    RESET_EMAIL_SENT,
    RESET_USER_PASSWORD_REQUEST,
    RESET_USER_PASSWORD_SUCCESS,
    RESET_USER_PASSWORD_ERROR,
    RESET_PASSWORD_RESET,
} from '../../constants/reset-password';

export interface ISendResetUserPasswordEmailAction {
    readonly type: typeof SEND_RESET_USER_PASSWORD_EMAIL_REQUEST;
}

export interface ISendResetUserPasswordEmailSuccessAction {
    readonly type: typeof SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS;
}

export interface ISendResetUserPasswordEmailFailedAction {
    readonly type: typeof SEND_RESET_USER_PASSWORD_EMAIL_ERROR;
    readonly message: string;
}

export interface IResetEmailSentAction {
    readonly type: typeof RESET_EMAIL_SENT;
}

export interface IResetUserPasswordAction {
    readonly type: typeof RESET_USER_PASSWORD_REQUEST;
}

export interface IResetUserPasswordSuccessAction {
    readonly type: typeof RESET_USER_PASSWORD_SUCCESS;
}

export interface IResetUserPasswordFailedAction {
    readonly type: typeof RESET_USER_PASSWORD_ERROR;
    readonly message: string;
}

export interface IResetPasswordResetAction {
    readonly type: typeof RESET_PASSWORD_RESET;
}

export type TResetPasswordActions =
    | ISendResetUserPasswordEmailAction
    | ISendResetUserPasswordEmailSuccessAction
    | ISendResetUserPasswordEmailFailedAction
    | IResetEmailSentAction
    | IResetUserPasswordAction
    | IResetUserPasswordSuccessAction
    | IResetUserPasswordFailedAction
    | IResetPasswordResetAction;
