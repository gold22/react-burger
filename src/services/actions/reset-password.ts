import {
    SEND_RESET_USER_PASSWORD_EMAIL_REQUEST,
    SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS,
    SEND_RESET_USER_PASSWORD_EMAIL_ERROR,
    RESET_EMAIL_SENT,
    RESET_USER_PASSWORD_REQUEST,
    RESET_USER_PASSWORD_SUCCESS,
    RESET_USER_PASSWORD_ERROR,
    RESET_PASSWORD_RESET,
} from '../constants/reset-password';
import { TDispatch, TGetState, TThunk } from '../types/store';
import { TApiUserResetPasswordInfo } from '../types/api';
import { IResetEmailSentAction, IResetPasswordResetAction } from '../types/actions/reset-password';
import ApiClient from '../api-client';
import { getErrorMessage } from '../../utils/error';

export const sendResetUserPasswordEmail: TThunk = (email: string) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: SEND_RESET_USER_PASSWORD_EMAIL_REQUEST,
    });
    try {
        await apiClient.sendResetUserPasswordEmail(email);
        dispatch({
            type: SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS,
        });
    } catch (error) {
        dispatch({
            type: SEND_RESET_USER_PASSWORD_EMAIL_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const resetEmailSent = (): IResetEmailSentAction => ({
    type: RESET_EMAIL_SENT,
});

export const resetUserPassword: TThunk = ({ password, token }: TApiUserResetPasswordInfo) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: RESET_USER_PASSWORD_REQUEST,
    });
    try {
        await apiClient.resetUserPassword({ password, token });
        dispatch({
            type: RESET_USER_PASSWORD_SUCCESS,
        });
    } catch (error) {
        dispatch({
            type: RESET_USER_PASSWORD_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const resetPasswordReset = (): IResetPasswordResetAction => ({
    type: RESET_PASSWORD_RESET,
});
