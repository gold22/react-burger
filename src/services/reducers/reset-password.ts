import {
    SEND_RESET_USER_PASSWORD_EMAIL_REQUEST,
    SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS,
    SEND_RESET_USER_PASSWORD_EMAIL_ERROR,
    RESET_EMAIL_SENT,
    RESET_USER_PASSWORD_REQUEST,
    RESET_USER_PASSWORD_SUCCESS,
    RESET_USER_PASSWORD_ERROR,
    RESET_PASSWORD_RESET,
} from '../constants/reset-password';
import { TResetPasswordActions } from '../types/actions/reset-password';

export type TResetPasswordState = {
    isEmailSent: boolean;
    isSendingEmail: boolean;
    sendingEmailError: string | null;
    isPasswordReset: boolean;
    isResettingPassword: boolean;
    resetPasswordError: string | null;
};

const initialState: TResetPasswordState = {
    isEmailSent: false,
    isSendingEmail: false,
    sendingEmailError: null,
    isPasswordReset: false,
    isResettingPassword: false,
    resetPasswordError: null,
};

export const resetPasswordReducer = (
    // eslint-disable-next-line @typescript-eslint/default-param-last
    state: TResetPasswordState = initialState,
    action: TResetPasswordActions,
) => {
    switch (action.type) {
    case SEND_RESET_USER_PASSWORD_EMAIL_REQUEST: {
        return {
            ...state,
            isSendingEmail: true,
        };
    }
    case SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS: {
        return {
            ...initialState,
            isEmailSent: true,
        };
    }
    case SEND_RESET_USER_PASSWORD_EMAIL_ERROR: {
        return {
            ...initialState,
            sendingEmailError: action.message,
        };
    }
    case RESET_EMAIL_SENT: {
        return {
            ...initialState,
        };
    }
    case RESET_USER_PASSWORD_REQUEST: {
        return {
            ...state,
            isResettingPassword: true,
        };
    }
    case RESET_USER_PASSWORD_SUCCESS: {
        return {
            ...initialState,
            isPasswordReset: true,
        };
    }
    case RESET_USER_PASSWORD_ERROR: {
        return {
            ...initialState,
            resetPasswordError: action.message,
        };
    }
    case RESET_PASSWORD_RESET: {
        return {
            ...initialState,
        };
    }
    default: {
        return state;
    }
    }
};
