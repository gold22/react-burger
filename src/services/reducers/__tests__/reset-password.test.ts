import { resetPasswordReducer, TResetPasswordState } from '../reset-password';
import {
    RESET_EMAIL_SENT,
    RESET_PASSWORD_RESET,
    RESET_USER_PASSWORD_ERROR,
    RESET_USER_PASSWORD_REQUEST,
    RESET_USER_PASSWORD_SUCCESS,
    SEND_RESET_USER_PASSWORD_EMAIL_ERROR,
    SEND_RESET_USER_PASSWORD_EMAIL_REQUEST,
    SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS,
} from '../../constants/reset-password';

const initialState: TResetPasswordState = resetPasswordReducer(undefined, {} as any);

describe('reset password reducer', () => {
    it('should return the initial state', () => {
        expect(initialState).toMatchSnapshot();
    });

    it('should handle SEND_RESET_USER_PASSWORD_EMAIL_REQUEST', () => {
        expect(resetPasswordReducer(undefined, { type: SEND_RESET_USER_PASSWORD_EMAIL_REQUEST }))
            .toMatchSnapshot();
    });

    it('should handle SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS', () => {
        const state: TResetPasswordState = {
            ...initialState,
            isSendingEmail: true,
        };
        expect(resetPasswordReducer(state, { type: SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS }))
            .toMatchSnapshot();
    });

    it('should handle SEND_RESET_USER_PASSWORD_EMAIL_ERROR', () => {
        const state: TResetPasswordState = {
            ...initialState,
            isSendingEmail: true,
        };
        expect(resetPasswordReducer(state, { type: SEND_RESET_USER_PASSWORD_EMAIL_ERROR, message: 'Send email error' }))
            .toMatchSnapshot();
    });

    it('should handle RESET_EMAIL_SENT', () => {
        const state: TResetPasswordState = {
            ...initialState,
            isEmailSent: true,
        };
        expect(resetPasswordReducer(state, { type: RESET_EMAIL_SENT }))
            .toMatchSnapshot();
    });

    it('should handle RESET_USER_PASSWORD_REQUEST', () => {
        expect(resetPasswordReducer(undefined, { type: RESET_USER_PASSWORD_REQUEST }))
            .toMatchSnapshot();
    });

    it('should handle RESET_USER_PASSWORD_SUCCESS', () => {
        const state: TResetPasswordState = {
            ...initialState,
            isResettingPassword: true,
        };
        expect(resetPasswordReducer(state, { type: RESET_USER_PASSWORD_SUCCESS }))
            .toMatchSnapshot();
    });

    it('should handle RESET_USER_PASSWORD_ERROR', () => {
        const state: TResetPasswordState = {
            ...initialState,
            isResettingPassword: true,
        };
        expect(resetPasswordReducer(state, { type: RESET_USER_PASSWORD_ERROR, message: 'Reset password error' }))
            .toMatchSnapshot();
    });

    it('should handle RESET_PASSWORD_RESET', () => {
        const state: TResetPasswordState = {
            ...initialState,
            isPasswordReset: true,
        };
        expect(resetPasswordReducer(state, { type: RESET_PASSWORD_RESET }))
            .toMatchSnapshot();
    });
});
