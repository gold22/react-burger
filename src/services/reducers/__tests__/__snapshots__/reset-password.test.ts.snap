// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reset password reducer should handle RESET_EMAIL_SENT 1`] = `
Object {
  "isEmailSent": false,
  "isPasswordReset": false,
  "isResettingPassword": false,
  "isSendingEmail": false,
  "resetPasswordError": null,
  "sendingEmailError": null,
}
`;

exports[`reset password reducer should handle RESET_PASSWORD_RESET 1`] = `
Object {
  "isEmailSent": false,
  "isPasswordReset": false,
  "isResettingPassword": false,
  "isSendingEmail": false,
  "resetPasswordError": null,
  "sendingEmailError": null,
}
`;

exports[`reset password reducer should handle RESET_USER_PASSWORD_ERROR 1`] = `
Object {
  "isEmailSent": false,
  "isPasswordReset": false,
  "isResettingPassword": false,
  "isSendingEmail": false,
  "resetPasswordError": "Reset password error",
  "sendingEmailError": null,
}
`;

exports[`reset password reducer should handle RESET_USER_PASSWORD_REQUEST 1`] = `
Object {
  "isEmailSent": false,
  "isPasswordReset": false,
  "isResettingPassword": true,
  "isSendingEmail": false,
  "resetPasswordError": null,
  "sendingEmailError": null,
}
`;

exports[`reset password reducer should handle RESET_USER_PASSWORD_SUCCESS 1`] = `
Object {
  "isEmailSent": false,
  "isPasswordReset": true,
  "isResettingPassword": false,
  "isSendingEmail": false,
  "resetPasswordError": null,
  "sendingEmailError": null,
}
`;

exports[`reset password reducer should handle SEND_RESET_USER_PASSWORD_EMAIL_ERROR 1`] = `
Object {
  "isEmailSent": false,
  "isPasswordReset": false,
  "isResettingPassword": false,
  "isSendingEmail": false,
  "resetPasswordError": null,
  "sendingEmailError": "Send email error",
}
`;

exports[`reset password reducer should handle SEND_RESET_USER_PASSWORD_EMAIL_REQUEST 1`] = `
Object {
  "isEmailSent": false,
  "isPasswordReset": false,
  "isResettingPassword": false,
  "isSendingEmail": true,
  "resetPasswordError": null,
  "sendingEmailError": null,
}
`;

exports[`reset password reducer should handle SEND_RESET_USER_PASSWORD_EMAIL_SUCCESS 1`] = `
Object {
  "isEmailSent": true,
  "isPasswordReset": false,
  "isResettingPassword": false,
  "isSendingEmail": false,
  "resetPasswordError": null,
  "sendingEmailError": null,
}
`;

exports[`reset password reducer should return the initial state 1`] = `
Object {
  "isEmailSent": false,
  "isPasswordReset": false,
  "isResettingPassword": false,
  "isSendingEmail": false,
  "resetPasswordError": null,
  "sendingEmailError": null,
}
`;
